### ISSUE TITLE
Cannot configure daemon's bip/bridge IP range

### ISSUE THREAD

**Issue by yacn** [docker/for-mac#218]
### Expected behavior

I should be able to configure the `bip` setting for the Docker Engine daemon running in xhyve via the Docker For Mac GUI. The default CIDR conflicts with our VPN's network range.
### Actual behavior

I cannot configure `bip` setting because there is no `etc/docker/daemon.json` to edit or option to change in the GUI. 
### Information
- Diagnostic ID: 30EC86B0-7ED8-44CD-B855-57327A022FBD
- OS X 10.10.5
### Steps to reproduce the behavior

I cannot from a blank slate configure the daemon's `bip` using the latest Docker For Mac stable or beta. To configure `bip`, I have to follow these steps:
1. Install Docker For Mac beta build 10258 (latest beta I have on my computer from before the stable/beta channel releases)
2. Go to config's directory: `cd ~/Library/Containers/com.docker.docker/Data/database/com.docker.driver.amd64-linux`
3. Update `daemon.json` to include `bip` config: `echo '{"storage-driver":"aufs","debug":"true","bip":"10.10.10.1/24"}' > etc/docker/daemon.json`
4. Add and commit changed `daemon.json`: `git add etc/docker/daemon.json && git commit -m "configure bip"`
5. After committing, xhyve is triggered to restart.
6. Run an alpine container and verify I can ping hosts within our VPN from the container
7. Quit Docker for Mac
8. Move Docker.app build 10258 out of Applications
9. Move the Docker for Mac (latest stable or beta release) into Applications
10. Start Docker for Mac, run alpine container and verify can still ping a host in the VPN

**Comment by yacn** [comment]
Just an update:

We can also retrieve the files by changing into `~/Library/Containers/com.docker.docker/Data/database` and performing `git reset --hard`. This puts the we need to modify back and allows us to change them as we see fit.

**Comment by algesten** [comment]
Docker 1.12.3. I can't get this to work.

I commit the change, quit and restart docker. Then it just stalls in "Docker is starting".

**Comment by hector70** [comment]
Docker 1.12.3 cannot find where to configure docker IP range (it's actually giving a network which is connected to my intranet causing conflicts).

**Comment by CerebralMastication** [comment]
I had the same issue as @algesten where docker would not restart after the change. I simplified the addition to daemon.json to simply {"bip":"172.17.0.1/24"} then I did the commit and things seemed to work for me.

**Comment by hector70** [comment]
Where can I find daemon.json on Mac docker

Il giorno lun 12 dic 2016 alle 22:21 JD Long <notifications@github.com> ha
scritto:

**Comment by CerebralMastication** [comment]
@hector70 the daemon.json is located in ~/Library/Containers/com.docker.docker/Data/database/com.docker.driver.amd64-linux/etc/docker

but note that the git repo is in ~/Library/Containers/com.docker.docker/Data/database

you'll have to do the  'git reset --hard' as discussed above then add the daemon.json to the repo then commit

**Comment by dsheets** [comment]
Since Beta 31, `daemon.json` is editable via :whale: ➡️ `Preferences...` ➡️ `Daemon` ➡️ `Advanced`.

**Comment by algesten** [comment]
Also it seems docker-compose have hard coded ranges that won't pick up on changed defaults in the docker `daemon.js`. So whilst single containers launched with `docker` get the IP change, networks created implicitly with `docker-compose` won't. Can be sorted in docker-compose.yml:

```
networks:
  default:
    driver: bridge
    ipam:
      driver: default
      config:
      - subnet:  10.103.0.1/16
        gateway: 10.103.0.1
```

**Comment by CerebralMastication** [comment]
@dsheets that will be great when that's in stable.

**Comment by samoht** [comment]
It is now possible to edit `daemon.json` in the UI on both the stable and the beta channels, so closing.

**Comment by BretFisher** [comment]
TL;DR for those that end up here like I did when having this problem

If you happen to have this problem where you're in a container and can't access your local network resources, and your local network is a 172.17, 172.18, 172.19, etc network, then you'll need to change the network docker is using so it doesn't conflict with your local LAN, VPN, etc.

bridge (docker0) defaults to 172.17.0.0/16

To change that, open up Docker for Mac Preferences and add a different private subnet to the Daemon config that you know won't conflict. I this example I picked a random 192.168 subnet:

<img width="464" alt="screen shot 2017-02-02 at 4 39 34 pm" src="https://cloud.githubusercontent.com/assets/792287/22570101/ff2e453a-e967-11e6-969a-9325571ad0c3.png">

Note this will only change the `bridge` network subnet, and not overlay's (used by docker-compose and swarm), or docker_gwbridge. In my experience, they layout like this:

bridge (docker0) : "Subnet": "172.17.0.0/16"
docker_gwbridge : "Subnet": "172.18.0.0/16"
any bridge/overlay networks you create after that: "Subnet": "172.19.0.0/16" and so on.

**Comment by AKFourSeven** [comment]
I really like this feature but it seems that other options are not working, for instance, I need to pass in these options :
```
DOCKER_OPTS = "--bip=xx.xx.xx.x/xx --dns xx.xx.x.xx --dns xx.xx.x.xx --dns-search xx.xx.xx"﻿
```

But when I try to add any of the dns or dns-search options in the json the restart goes idle meaning that the added options are breaking the config. 
How then would I go on adding these options ?
Can anyone help at all ?
I am on Docker CE - Version 17.03.1-ce-mac12 (17661) running on macOS Sierra.

**Comment by willthames** [comment]
I had a few problems with the `bip` setup - it's not obvious that `bip` is both an IP address and a subnet mask at the same time. If you set it to a subnet mask, Docker won't start and you'll have to reset to factory defaults.

Good:

```
{
  "bip": "10.10.0.1/16"
}
```

Bad:
```
{ 
  "bip": "10.10.0.0/16"
}
```

**Comment by CelsoSantos** [comment]
The true question for me is, setting the `bip` value will allow me to interact/connect Vagrant VMs to Docker? I don't think this is enough. Can anyone confirm please?

Since not using docker-toolbox things seem to have gotten a lot more complicated to achieve the intended scenario..

**Comment by willthames** [comment]
@CelsoSantos the only reason I set the `bip` is to avoid a clash with an internal network on the same address range - unless vagrant also runs on that subnet range I don't think setting `bip` will help. I don't know what will though, sorry :(
